airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/command/DestinationCatalog.kt

```diff
@@ -29,6 +29,18 @@ interface DestinationCatalogFactory {
     fun make(): DestinationCatalog
 }
 
+// I'd caution against using Micronaut for this.
+// and instead consider inlining the generation of the DestinationCatalog in the WriteOperation
+// and then plumb it and its friends in some immutable DestinationContext data class.
+//
+// The main reason for this is that DI quickly becomes tentacular and hard to reason about.
+// In other words, there's a very real complexity cost to every new injectable bean that you add.
+// It adds up very quickly and is super hard to simplify later on.
+//
+// Furthermore, you open yourself up to exceptions being thrown at startup time when the
+// factory does its thing, which is probably not what you want: if 1 out of 10 streams is bad you
+// probably want to print some stream error message for the bad stream but keep chugging along
+// for the 9 others.
```

Comment:
This strikes me as exactly what dependency injection is for? We decision on the catalog in multiple places, the simplest thing to do is to inject it everywhere it's needed.

Reply:
Sorry, that's not what I meant. 

Suppose you're injecting a `DestinationCatalog`. Instantiating this singleton will require a `ConfiguredAirbyteCatalog`. Suppose this singleton was instantiated and is superficially valid but one of the streams is corrupted somehow and causes the `DestinationStreamFactory` to throw an exception. 

To deal with this very real possibility, you now have to concern yourself with exception handling; which is doable but annoying because it either happens at startup or if you're lazy (with a `Provider<>`) at some point in time that you don't control.

If you create this singleton the old fashioned way, you control when instantiation happens. It's a tradeoff, like yeah if plumbing your artisanally hand-crafted singleton is a pain then yeah DI is undoubtedly the better choice, but otherwise if it's no big deal then in my opinion DI shouldn't be the default choice.